dal/elastic.go

```diff
@@ -2,126 +2,109 @@ package dal
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
+	"github.com/YuvalSteiy/book_service/models"
 	"github.com/olivere/elastic/v7"
-	"log"
-	"strconv"
 )
-const INDEXNAME = "books_yuval"
-// Book struct for deserialization of documents from json
-type Book struct {
-	Title          string  `json:"title"`
-	AuthorName     string  `json:"author_name"`
-	EbookAvailable bool    `json:"ebook_available"`
-	Price          float64 `json:"price"`
-	PublishDate    string  `json:"publish_date"`
+
+const INDEX_NAME = "books_yuval"
+const PORT = 9200
+const DOMAIN_URL = "http://es-search-7.fiverrdev.com"
+
+type Elastic struct {
+	Client *elastic.Client
 }
 
-func ElasticInitClient() *elastic.Client {
-	// Pass connection port
-	port := 9200
-	strPort := strconv.Itoa(port)
-	// Pass domain string
-	domainURL := "http://es-search-7.fiverrdev.com"
-	// Complete URL
-	setURL := domainURL + ":" + strPort
-	// Instantiate a client (elastic library instance)
+func (e *Elastic) InitClient() error {
+	setURL := fmt.Sprint(DOMAIN_URL, ":", PORT)
 	client, err := elastic.NewClient(elastic.SetURL(setURL))
 	if err != nil {
-		fmt.Println("elastic.NewClient() ERROR:", err)
-		log.Fatalf("quitting connection..")
+		return err
 	}
-	return client
+	e.Client = client
+	return nil
 }
 
-func ElasticGetBookByID(client *elastic.Client, id string) *elastic.GetResult{
+func (e *Elastic) GetBookByID(id string) (*models.Book, error) {
 	ctx := context.Background()
-	get1, err := client.Get().Index(INDEXNAME).Id(id).Do(ctx)
+	book, err := (e.Client).Get().Index(INDEX_NAME).Id(id).Do(ctx)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
-	return get1
+	var parsedBook models.Book
+	json.Unmarshal(book.Source, &parsedBook)
+	return &parsedBook, nil
 }
 
-func ElasticGetStoreData(client *elastic.Client) (int64, *elastic.SearchResult){
+func (e *Elastic) InsertBook(book *models.Book) (string, error) {
 	ctx := context.Background()
-	//get document count in index
-	countService := elastic.NewCountService(client)
-	countResult, err := countService.Index(INDEXNAME).Do(ctx)
+	putStatus, err := (e.Client).Index().Index(INDEX_NAME).BodyJson(*book).Do(ctx)
```

Comment:
why do you get a pointer if you need the object itself? and if you need to receive it as a pointer - please check that it is not nil before you are performing `*book`